libkvs: cleanup, deprecate old functions, update users

garlick opened the pull request:
This PR is an attempt at some long overdue KVS API cleanup.  Not ready for merge consideration until a PR is submitted to update sched.

All KVS API components are prefixed with `flux_` now, and the `kvs_*`, `kvsitr_*`, and `kvsdir_*` prefixes are dropped from flux-core symbol exports.

The older API functions defined in kvs_classic.h are marked with `__attribute__ ((deprecated))`.  I stopped short of moving these functions to libcompat, which would take them out of the exported namespace, due to difficulties with the python bindings.  The old "typed" get/put functions that remain were replaced with jansson pack/unpack style equivalents to reduce the deprecated API footprint somewhat.

The kvs_dir.h functions, while still not quite right,  were cleaned up a bit, including making some `flux_kvsdir_t *` parameters  `const` that ought to have been.

Some new functions were introduced:
```c
flux_kvsdir_copy()
flux_kvs_lookup_get_vunpack()
flux_kvs_lookup_get_dir()
flux_kvs_txn_vpack()
```

Finally, inline (header) documentation was added/updated for kvs_dir.h, kvs_watch.h, and kvs_classic.h which are still not quite stable IMHO.  Man pages were added for the new lookup and txn functions.

I did not fully modernize the API usage of wreck, libkz, and the lua bindings since some of that code is going to be replaced and it was pretty involved.  There will be a few more deprecated function warnings when building those for now.

The kvs_watch.h API is still the old one and in dire need of redesign, but not dealt with here.

garlick commented:
Hmm, getting some travis failures and I'm having a hard time spotting the root cause.  Best I can tell is things are going wrong with distcheck in the python bindings...

```
make[4]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python'
Making check in flux
make[5]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
Making check in core
make[6]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux/core'
echo Copying core libraries to where they can be used by python out of tree
Copying core libraries to where they can be used by python out of tree
[ "../../../../../../.." != "../../../../.." ] && cp  ../../../../../../../src/bindings/python/flux/core/*.py ./ || true
cp: cannot create regular file ‘./handle.py’: Permission denied
cp: cannot create regular file ‘./__init__.py’: Permission denied
cp: cannot create regular file ‘./inner.py’: Permission denied
cp: cannot create regular file ‘./trampoline.py’: Permission denied
cp: cannot create regular file ‘./watchers.py’: Permission denied
make  check-local
make[7]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux/core'
echo Copying core libraries to where they can be used by python out of tree
Copying core libraries to where they can be used by python out of tree
[ "../../../../../../.." != "../../../../.." ] && cp  ../../../../../../../src/bindings/python/flux/core/*.py ./ || true
cp: cannot create regular file ‘./handle.py’: Permission denied
cp: cannot create regular file ‘./__init__.py’: Permission denied
cp: cannot create regular file ‘./inner.py’: Permission denied
cp: cannot create regular file ‘./trampoline.py’: Permission denied
cp: cannot create regular file ‘./watchers.py’: Permission denied
make[7]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux/core'
make[6]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux/core'
make[6]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
echo Copying libraries to where they can be used by python in-tree
Copying libraries to where they can be used by python in-tree
for LIB in _core.so _kvs.so _jsc.so _kz.so ; do \
	    cp -f .libs/$LIB ./ ; \
	done
[ "../../../../../.." != "../../../.." ] && cp ../../../../../../src/bindings/python/flux/*.py ./ || true
cp: cannot create regular file ‘./constants.py’: Permission denied
cp: cannot create regular file ‘./__init__.py’: Permission denied
cp: cannot create regular file ‘./jsc.py’: Permission denied
cp: cannot create regular file ‘./kvs.py’: Permission denied
cp: cannot create regular file ‘./kz.py’: Permission denied
cp: cannot create regular file ‘./message.py’: Permission denied
cp: cannot create regular file ‘./rpc.py’: Permission denied
cp: cannot create regular file ‘./sec.py’: Permission denied
cp: cannot create regular file ‘./wrapper.py’: Permission denied
make  check-local
make[7]: Entering directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
echo Copying libraries to where they can be used by python in-tree
Copying libraries to where they can be used by python in-tree
for LIB in _core.so _kvs.so _jsc.so _kz.so ; do \
	    cp -f .libs/$LIB ./ ; \
	done
[ "../../../../../.." != "../../../.." ] && cp ../../../../../../src/bindings/python/flux/*.py ./ || true
cp: cannot create regular file ‘./constants.py’: Permission denied
cp: cannot create regular file ‘./__init__.py’: Permission denied
cp: cannot create regular file ‘./jsc.py’: Permission denied
cp: cannot create regular file ‘./kvs.py’: Permission denied
cp: cannot create regular file ‘./kz.py’: Permission denied
cp: cannot create regular file ‘./message.py’: Permission denied
cp: cannot create regular file ‘./rpc.py’: Permission denied
cp: cannot create regular file ‘./sec.py’: Permission denied
cp: cannot create regular file ‘./wrapper.py’: Permission denied
if [ -x "$( which pylint )" ] ; then  pylint --rcfile=../../../../../../src/bindings/python/.pylintrc ../flux ; fi
Warning: option profile is obsolete and it is slated for removal in Pylint 1.6.
Warning: option include-ids is obsolete and it is slated for removal in Pylint 1.6.
Warning: option symbols is obsolete and it is slated for removal in Pylint 1.6.
Warning: option comment is obsolete and it is slated for removal in Pylint 1.6.
Warning: option zope is obsolete and it is slated for removal in Pylint 1.6.
************* Module flux
I:  5, 0: Locally disabling wildcard-import (W0401) (locally-disabled)
************* Module flux._core_build
I:  2, 0: Ignoring entire file (file-ignored)
************* Module flux._kz_build
I:  2, 0: Ignoring entire file (file-ignored)
************* Module flux._kvs_build
I:  2, 0: Ignoring entire file (file-ignored)
************* Module flux.kvs
I: 97, 0: Locally disabling too-many-ancestors (R0901) (locally-disabled)
I: 97, 0: Locally disabling too-many-public-methods (R0904) (locally-disabled)
C:308, 0: Line too long (88/80) (line-too-long)
************* Module flux._jsc_build
I:  2, 0: Ignoring entire file (file-ignored)
************* Module flux.core.handle
I: 88, 0: Locally disabling no-self-use (R0201) (locally-disabled)
************* Module flux.core.inner
I: 20, 0: Locally disabling invalid-name (C0103) (locally-disabled)
Report
======
730 statements analysed.
Statistics by type
------------------
+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |14     |NC         |NC         |21.43       |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |27     |NC         |NC         |22.22       |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |106    |NC         |NC         |70.75       |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|function |30     |NC         |NC         |6.67        |0.00     |
+---------+-------+-----------+-----------+------------+---------+
External dependencies
---------------------
::
    flux 
      \-_core (flux.core.handle,flux.constants,flux.core.inner,flux.sec,flux.core.trampoline)
      \-_jsc (flux.jsc)
      \-_kvs (flux.kvs)
      \-_kz (flux.kz)
      \-constants (flux.message,flux.rpc)
      \-core (flux)
      | \-handle (flux.core,flux.core.trampoline)
      | \-inner (flux.message,flux.core.watchers,flux.core.handle,flux.rpc)
      | \-trampoline (flux.core)
      | \-watchers (flux.message,flux.core.handle)
      \-message (flux.core.handle)
      \-rpc (flux.core.handle)
      \-wrapper (flux.kvs,flux.message,flux.jsc,flux.sec,flux.rpc,flux.kz,flux.core.inner,flux.core.handle)
Raw metrics
-----------
+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |963    |68.98 |NC       |NC         |
+----------+-------+------+---------+-----------+
|docstring |127    |9.10  |NC       |NC         |
+----------+-------+------+---------+-----------+
|comment   |45     |3.22  |NC       |NC         |
+----------+-------+------+---------+-----------+
|empty     |261    |18.70 |NC       |NC         |
+----------+-------+------+---------+-----------+
Duplication
-----------
+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |NC       |NC         |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |NC       |NC         |
+-------------------------+------+---------+-----------+
Messages by category
--------------------
+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |1      |NC       |NC         |
+-----------+-------+---------+-----------+
|refactor   |0      |NC       |NC         |
+-----------+-------+---------+-----------+
|warning    |0      |NC       |NC         |
+-----------+-------+---------+-----------+
|error      |0      |NC       |NC         |
+-----------+-------+---------+-----------+
% errors / warnings by module
-----------------------------
+---------+------+--------+---------+-----------+
|module   |error |warning |refactor |convention |
+=========+======+========+=========+===========+
|flux.kvs |0.00  |0.00    |0.00     |100.00     |
+---------+------+--------+---------+-----------+
Messages
--------
+-----------------+------------+
|message id       |occurrences |
+=================+============+
|locally-disabled |5           |
+-----------------+------------+
|file-ignored     |4           |
+-----------------+------------+
|line-too-long    |1           |
+-----------------+------------+
Global evaluation
-----------------
Your code has been rated at 9.99/10
make[7]: *** [check-local] Error 16
make[7]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
make[6]: *** [check-am] Error 2
make[6]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
make[5]: *** [check-recursive] Error 1
make[5]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python/flux'
make[4]: *** [check-recursive] Error 1
make[4]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings/python'
make[3]: *** [check-recursive] Error 1
make[3]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src/bindings'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub/src'
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/flux-framework/flux-core/flux-core-0.8.0-230-g8cb94f0d/_build/sub'
make: *** [distcheck] Error 
```

garlick commented:
Pylint problem solved.

Still todo: add some tests to cover missing bits, and put together a patch to update sched.

chu11 commented on src/common/libkvs/kvs_dir.c:
```diff
@@ -106,33 +109,27 @@ flux_kvsdir_t *flux_kvsdir_create (flux_t *handle, const char *rootref,
-            return NULL;
-        }
-    }
-    return dir->dirobj_string;
```
should the change from using ```dirobj_string``` to ```json_str``` be in a different commit, b/c this doesn't seem to be mapped to this commit message.  It appears this also fixes a mem-leak.

garlick replied:
Yes that looks like a rebase error.  I'll fix...

chu11 commented on src/common/libkvs/kvs_lookup.c:
```diff
@@ -37,35 +37,48 @@
 {
     if (ctx) {
+        free (ctx->key);
         free (ctx->treeobj_str);
```
should we free ```atref``` too?  Can't see where else it is freed.

garlick replied:
Yes good catch!  I'll fix that.

chu11 commented on src/bindings/lua/flux-lua.c:
```diff
@@ -254,8 +254,11 @@ static int l_flux_new (lua_State *L)
 static int l_flux_kvsdir_new (lua_State *L)
```
commit seems to do more than just update to use ```flux_kvs_lookup_get_dir```.  Perhaps commit message should just be enhanced.

chu11 replied:
oh yeah, this specific line is not relevant, just didn't know how to commit on a commit message.

garlick replied:
Yes that was pretty vague.  Changed to
>   bindings/lua: use flux_kvs_lookup()
> 
>  Replace use of soon to be deprecated flux_kvs_get()
    and flux_kvs_get_dir() with modern flux_kvs_lookup()
    equivalents.

chu11 commented on src/common/libkvs/kvs_watch.h:
```diff
@@ -5,45 +5,76 @@
  */
 int flux_kvs_watch_once (flux_t *h, const char *key, char **json_str);
+
+/* Same as above except value is a direcory pointed to by 'dirp'.
```
typo ```directory```

chu11 replied:
raced on comments/old commit and github is showing my comment as outdated.  typo ```directory```

chu11 commented:
just finished skimming over everything.  Overall looks good.  Lots of lines of code, so hopefully I didn't miss anything.  I think @grondo should look over the wreck and job areas that were updated, as I don't know that code as well.

garlick commented:
Thanks - fixed the "directory" typo, squashed and pushed.

grondo commented on doc/man3/flux_kvs_lookup.adoc:
```diff
@@ -53,6 +55,9 @@ the returned JSON is parsed according to variable arguments in Jansson
 
+`flux_kvs_lookup_get_dir()` is identical to `flux_kvs_lookup_get()` except
+a directory object is returned.
+
```
I know you are considering removal of the `kvsdir` abstraction, but in the context of the addition of `flux_kvs_lookup_get_dir` it might be useful to consider addition of a refcount to kvsdir objects. This would avoid the necessary copies in the common case where we want the kvsdir to persist, but we're done with the future.

garlick replied:
We do have `flux_kvsdir_incref()` but since the object returned from the lookup is const (indicating that it still belongs to the future),  it seems a bit wrong to allow its reference count to be modified?  This is why I added `flux_kvsdir_copy()`, but I intended that to be short lived until the various uses where `flux_kvsdir_t` was used purposes other than "list of names in directory" were converted to something else TBD.

As we just discussed offline, possibly we need to come up with a new object that serves as a handle to the KVS, and contains things like a "current working directory", a snapshot reference, or a namespace handle, as well as the broker handle, and then all KVS functions would take one of those instead of the broker handle...  I'll open an issue on that idea.

grondo commented on src/modules/wreck/wrexecd.c:
```diff
@@ -1651,20 +1655,29 @@ static int l_push_environ (lua_State *L, int index)
+                    || !(f = flux_kvs_lookup (h, FLUX_KVS_READDIR, key))
+                    || flux_kvs_lookup_get_dir (f, &dir) < 0
+                    || !(t->kvs = flux_kvsdir_copy (dir))) {
+            if (errno != ENOENT)
```
this chunk of code is repeated twice at least in this source file, maybe it could be its own static function? In general, the lookup->get_dir->copy_dir seems to be common enough idiom that I'm wondering why there isn't a function in the API for it, but it isn't quite common enough for me to care (especially if the flux_kvsdir_t is going to go away.)

grondo commented:
One other thing. Looks like `flux_kvs_pack` isn't covered in testing, and that it is only necessary due to untested wrexecd case where `no-aggregate-task-exit` option is used.

Maybe we can just drop that part of wrexecd and remove `flux_kvs_pack`? If there is another perceived need for `flux_kvs_pack`, then should it have a unit test?

garlick commented:
> One other thing. Looks like flux_kvs_pack isn't covered in testing, and that it is only necessary due to untested wrexecd case where no-aggregate-task-exit option is used.

> Maybe we can just drop that part of wrexecd and remove flux_kvs_pack? If there is another perceived need for flux_kvs_pack, then should it have a unit test?

I'm good with that - let me see if I can do that really quick.  o/w I will add a test.

garlick commented:
Dropped `flux wreckrun --no-aggregate-task-exit` as discussed, and eliminated `flux_kvs_pack()` from the set of deprecated functions.  Squashed that down and re-pushed.
